src/parser/elements_parser.ts

```diff
@@ -1,92 +1,136 @@
+import {FabricObject} from "../shapes/object.class";
+
+export interface ElementsParserOptions extends TParsedViewBoxDims {
+  svgUid?: number;
+}
+
+export interface ElementsParserParsingOptions extends ElementsParserOptions {
+  /**
+   * see https://developer.mozilla.org/en-US/docs/Web/API/AbortController/signal
+   */
+  signal?: AbortSignal;
+  /**
+   * crossOrigin settings
+   */
+  crossOrigin?: string;
+}
+
+export type TElementsParserCallback = (
+  obj: FabricObject[],
+  container?: Element[]
+) => void;
+
+export type TReviver = (el: Element, obj: FabricObject) => void;
+
+class ElementsParser {
+  /**
+   * The elements of the SVG
+   */
+  elements: Element[];
+  callback: TElementsParserCallback;
+  options: ElementsParserOptions;
+  reviver: TReviver;
+  svgUid: number;
+  parsingOptions: ElementsParserParsingOptions;
+  regexUrl: RegExp;
+  doc?: Document;
 
-const ElementsParser = function (
-  elements,
-  callback,
-  options,
-  reviver,
-  parsingOptions,
-  doc
-) {
-  this.elements = elements;
-  this.callback = callback;
-  this.options = options;
-  this.reviver = reviver;
-  this.svgUid = (options && options.svgUid) || 0;
-  this.parsingOptions = parsingOptions;
-  this.regexUrl = /^url\(['"]?#([^'"]+)['"]?\)/g;
-  this.doc = doc;
-};
-
-(function (proto) {
-  proto.parse = function () {
+  instances?: Array<FabricObject>;
+  numElements?: number;
+
+  constructor(
+    elements: Element[],
+    callback: TElementsParserCallback,
+    options: ElementsParserOptions,
+    reviver: TReviver,
+    parsingOptions: ElementsParserParsingOptions,
+    doc?: Document
+  ) {
+    this.elements = elements;
+    this.callback = callback;
+    this.options = options;
+    this.reviver = reviver;
+    this.svgUid = (options && options.svgUid) || 0;
+    this.parsingOptions = parsingOptions;
+    this.regexUrl = /^url\(['"]?#([^'"]+)['"]?\)/g;
+    this.doc = doc;
+  }
+
+  parse() {
     this.instances = new Array(this.elements.length);
     this.numElements = this.elements.length;
     this.createObjects();
-  };
+  }
 
-  proto.createObjects = function () {
+  createObjects() {
     this.elements.forEach((element, i) => {
-      element.setAttribute('svgUid', this.svgUid);
+      element.setAttribute('svgUid', this.svgUid.toString());
       this.createObject(element, i);
     });
-  };
+  }
 
-  proto.findTag = function (el) {
+  findTag(el: Element): FabricObject {
     return fabric[capitalize(el.tagName.replace('svg:', ''))];
-  };
+  }
 
-  proto.createObject = function (el, index) {
+  createObject(el: Element, index: number) {
     const klass = this.findTag(el);
-    if (klass && klass.fromElement) {
+    // TODO: firgure out a better way. maybe modify sig of FabricObject?
+    if (klass && (klass as any).fromElement) {
       try {
-        klass.fromElement(el, this.createCallback(index, el), this.options);
+        (klass as any).fromElement(el, this.createCallback(index, el), this.options);
```

Comment:
fromElement exists on SOME fabric objects (like the gradient). Maybe we can add a default inherited method, IRDK.

Reply:
For these kinds of things best practice is to declare an interface and force classes to implement it with `implements`.
The stricter approach is declaring an abstract base class but that is most likely incompatible for fabric since Object extends a base class that is unrelated.